### PR TITLE
feat: property based tests for liquidity mining

### PR DESCRIPTION
**Pull request by martinfridrich**
Added property based test for liquidity mining

### CODE REVIEW

File: src/liquidity_mining/invariants.rs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+        assert!(multiplier < FixedU128::one(), "loyalty multiplier is >= 1");
+
+        //b + (1 - b)/2;
+        let bound = FixedU128::one()
+                .checked_sub(&b).unwrap()
+                .checked_div(&FixedU128::from(2_u128)).unwrap()
+                .checked_add(&b).unwrap();
+
+        if periods < scale_coef {
```
Comment (poliwop) [inline src/liquidity_mining/invariants.rs]: It's probably better to split into a test function where you guarantee periods <= scale_coef and a second test function where periods >= scale_coef

File: src/liquidity_mining/invariants.rs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            "loyalty multiplier, LoyaltyFactor * (t + tb + scaleCoef*(b + 1)) == t + tb + b*scaleCoef*(b + 1)"
+        );
+    }
+
+    #[test]
+    fn calculate_loyalty_multiplier_should_eqaul_to_bound_when_period_is_equal_scale_coef(
+        periods in periods(),
+        initial_reward_percentage in initial_reward_percentage(),
+    ) {
+        //This case never hit by test above, that's why there is special test only for this
```
Comment (poliwop) [inline src/liquidity_mining/invariants.rs]: Definitely agree with splitting this out as a separate test, it validates the meaning of the scale_coef parameter

**Review by poliwop — COMMENTED** [review]
Looks basically good.

On the question of order of magnitude of values to property test with... What I tend to do is test for some big range in the middle of expected values for variables, and then think about what else I want to test... With stableswap for example I'm carefully testing rounding and convergence, and I different tests narrowly restricting the variable ranges to some extreme values (some extremely high, some extremely low) that I know will stress whatever I'm trying to test. Mostly this is useful to test convergence of the numerical method we are using in Stableswap. I do not immediately see how it would be useful here, but you are more familiar with the liquidity mining and you may think of some way/reason to test extreme values.

TLDR is that fuzz testing doesn't do a great job of testing corner cases so you may want to think through some of those too, and then you can fuzz test specific narrow ranges of values as additional property tests. But I don't see any obvious test to add right now.
